File: src/opendms/search_index/tests/factories.py
import factory

from ..index import Document, Zaak


class IndexZaakFactory(factory.Factory):
    uuid = factory.Faker("uuid4", cast_to=str)
    url = factory.Faker("url")
    identificatie = factory.Faker("uuid4", cast_to=str)
    bronorganisatie = factory.Faker("company")
    verantwoordelijkeOrganisatie = factory.Faker("company")
    omschrijving = factory.Faker("sentence", nb_words=6)
    toelichting = factory.Faker("paragraph")
    status = factory.Faker("word")
    registratiedatum = factory.Faker("past_date")
    startdatum = factory.Faker("past_date")
    zaaktype = factory.Faker("word")

    service_slug = "zaken-api"
    ztc_service_slug = "catalogi-api"
    group_slug = "group-1"

    class Meta:
        model = Zaak


class NestedzaakFactory(factory.Factory):
    uuid = factory.Faker("uuid4", cast_to=str)
    url = factory.Faker("url")
    identificatie = factory.Faker("uuid4", cast_to=str)
    bronorganisatie = factory.Faker("company")
    verantwoordelijkeOrganisatie = factory.Faker("company")
    omschrijving = factory.Faker("sentence", nb_words=6)
    toelichting = factory.Faker("paragraph")
    status = factory.Faker("word")
    registratiedatum = factory.Faker("past_date")
    startdatum = factory.Faker("past_date")
    zaaktype = factory.Faker("word")
    object_type = factory.Faker("word")

    class Meta:
        model = dict


class IndexDocumentFactory(factory.Factory):
    uuid = factory.Faker("uuid4", cast_to=str)
    url = factory.Faker("url")
    identificatie = factory.Faker("uuid4", cast_to=str)
    bronorganisatie = factory.Faker("company")
    titel = factory.Faker("sentence", nb_words=6)
    beschrijving = factory.Faker("paragraph")
    auteur = factory.Faker("name")
    taal = factory.Faker("language_code")
    vertrouwelijkheidaanduiding = factory.Faker("word")
    status = factory.Faker("word")
    formaat = factory.Faker("file_extension")
    bestandsnaam = factory.Faker("file_name")
    informatieobjecttype = factory.Faker("word")
    verschijningsvorm = factory.Faker("word")
    bestandsomvang = 1000000
    link = factory.Faker("url")
    creatiedatum = factory.Faker("past_date")
    begin_registratie = factory.Faker("past_datetime")

    zaak_referenties = factory.List([factory.SubFactory(NestedzaakFactory)])

    class Meta:
        model = Document

File: src/opendms/frontend/.storybook/factories.ts
import type { Document, StatusType, Zaak } from "~/types";

type FactoryFn<T extends object> = (values?: Partial<T>) => T;

export const documentFactory = createFactory<Document>({
  hasPendingUpdates: false,
  identificatie: "doc-{index}",
  titel: "Document {index}",
  bestandsnaam: "document-{index}.pdf",
  beschrijving: null,
  formaat: "pdf",
  link: null,
  auteur: "",
  beginRegistratie: "",
  bronorganisatie: "",
  creatiedatum: "",
  informatieobjecttype: "",
  inhoud: "",
  taal: "",
  url: "",
  uuid: "",
});

export const statustypeFactory = createFactory<StatusType>({
  uuid: "00000000-0000-0000-0000-000000000000",
  url: "http://www.example.com/api/v1/services/catalogi-api/statustypen/00000000-0000-0000-0000-000000000000",
  catalogus: "",
  informeren: false,
  isEindstatus: false,
  omschrijving: "Sed do eiusmod",
  volgnummer: 0,
  zaaktype: "",
  zaaktypeIdentificatie: "",
});

export const zaakFactory = createFactory<Zaak>({
  identificatie: "ZAAK-2026-0000000010",
  omschrijving: "Lorem ipsum dolor sit amet",
  uuid: "00000000-0000-0000-0000-000000000000",
  bronorganisatie: "",
  registratiedatum: "",
  startdatum: "",
  toelichting: "",
  url: "http://www.example.com/zaken/api/v1/zaken/00000000-0000-0000-0000-000000000000",
  verantwoordelijkeOrganisatie: "",
  zaaktype:
    "http://www.example.com/catalogi/api/v1/zaaktypen/22222222-2222-2222-2222-222222222222",
});

/**
 * Creates a factory function for `<T>`.
 * @param fixture The base fixture to return when calling the produced factory function.
 * @return {FactoryFn} A function that takes `values` and assigns it to a cloned `fixture` (supplied by `createFactory`)
 * before returning it.
 */
function createFactory<T extends object>(fixture: T): FactoryFn<T> {
  return (values) => baseFactory(fixture, values);
}

/**
 * Creates `amount` instances of `<T>` using provided `factoryFn`.
 * @param {FactoryFn} factoryFn A function that takes `values` and assigns it to a cloned `fixture` (supplied by
 * `createFactory`) before returning it.
 * @param {number} amount
 * @param {Partial<T>|Partial<T>[]} values A (subset of) `<T>` object.
 */
export function batchFactory<T extends object>(
  factoryFn: FactoryFn<T>,
  amount: number,
  values?: Partial<T> | Partial<T>[],
): T[] {
  const valuesArray = Array.from({
    ...(Array.isArray(values) ? values : []),
    length: amount,
  });
  return valuesArray.map((_values, index) => {
    const object = factoryFn(_values || values);

    return Object.fromEntries(
      Object.entries(object).map(([key, value]) => {
        const newValue =
          typeof value === "string"
            ? value.replaceAll("{index}", (index + 1).toString())
            : value;
        return [key, newValue];
      }),
    ) as T;
  });
}

/**
 * The base factory function, creates an instance of `<T>` with values assigns to it.
 * Takes `values` and assigns it to a cloned `fixture` before returning it.
 * @private Don't use this directly, use `fooFactory = createFactory<Foo>({'foo': 'bar'})` instead.
 * @param fixture
 * @param values
 */
export function baseFactory<T extends object>(
  fixture: T,
  values?: Partial<T>,
): T {
  const clone = { ...fixture }; // Shallow clone for now.
  return Object.assign(clone, values || {});
}
